`@deno/kv` does not work when NAPI is disabled, even when running in Deno

### 🔍

- [X] Did you search for existing issues?

### Type of feedback

Bug report

### Description

Project linked to GitHub
Downloaded file:///src/main.ts (1/1)
Packaging complete
The deployment failed: UNCAUGHT_EXCEPTION Error: Cannot find module '@deno/kv-linux-x64-gnu' Require stack: - /node_modules/.deno/@deno+kv@0.8.3/node_modules/@deno/kv/esm/_napi_index.cjs - /node_modules/.deno/@deno+kv@0.8.3/node_modules/@deno/kv/esm/_napi_index.cjs at Function.Module._resolveFilename (node:module:608:15) at Function.Module._load (node:module:486:27) at Module.require (node:module:674:19) at require (node:module:801:16) at Object.<anonymous> (file:///node_modules/.deno/@deno+kv@0.8.3/node_modules/@deno/kv/esm/_napi_index.cjs:92:31) at Object.<anonymous> (file:///node_modules/.deno/@deno+kv@0.8.3/node_modules/@deno/kv/esm/_napi_index.cjs:126:4) at Module._compile (node:module:736:34) at Object.Module._extensions..js (node:module:757:11) at Module.load (node:module:655:32) at Function.Module._load (node:module:523:13)

### Steps to reproduce (if applicable)

The local test is fine, but there is an error with the deno cloud

### Expected behavior (if applicable)

_No response_

### Possible solution (if applicable)

_No response_

### Additional context

_No response_

--- a/npm/src/napi_based.ts
+++ b/npm/src/napi_based.ts
@@ -123,6 +123,7 @@ export function isNapiInterface(obj: unknown): obj is NapiInterface {
 
 //
 
+let DEFAULT_NAPI_INTERFACE_LOAD_ERROR: unknown;
 // deno-lint-ignore no-explicit-any
 const DEFAULT_NAPI_INTERFACE: any = undefined;
 
@@ -165,9 +166,14 @@ class NapiBasedKv extends ProtoBasedKv {
       throw new Error(`Invalid path: ${url}`);
     }
     if (napi === undefined) {
-      throw new Error(
+      const error = new Error(
         `No default napi interface, provide one via the 'napi' option.`,
       );
+      if (DEFAULT_NAPI_INTERFACE_LOAD_ERROR !== undefined) {
+        (error as { cause?: unknown }).cause =
+          DEFAULT_NAPI_INTERFACE_LOAD_ERROR;
+      }
+      throw error;
     }
     const dbId = napi.open(url, inMemory, debug);
     return new NapiBasedKv(debug, napi, dbId, decodeV8, encodeV8);

--- a/npm/src/scripts/build_npm.ts
+++ b/npm/src/scripts/build_npm.ts
@@ -109,9 +109,27 @@ await build({
       const oldContents = await Deno.readTextFile(
         join(outDir, subdir, "napi_based.js"),
       );
+      // Load the napi binding lazily-failing: platforms without a
+      // binding (e.g. Deno Deploy) must still be able to import the
+      // package and use the non-napi implementations. The load error
+      // is kept and surfaced only when the napi path is actually used.
       const insertion = subdir === "esm"
-        ? `await import('./${name}')`
-        : `require('./${name}')`;
+        ? `await (async () => {
+  try {
+    return await import('./${name}');
+  } catch (e) {
+    DEFAULT_NAPI_INTERFACE_LOAD_ERROR = e;
+    return undefined;
+  }
+})()`
+        : `(() => {
+  try {
+    return require('./${name}');
+  } catch (e) {
+    DEFAULT_NAPI_INTERFACE_LOAD_ERROR = e;
+    return undefined;
+  }
+})()`;
 
       const newContents = oldContents.replace(
         `const DEFAULT_NAPI_INTERFACE = undefined;`,

--- a/npm/src/scripts/build_npm.ts
+++ b/npm/src/scripts/build_npm.ts
@@ -140,6 +140,40 @@ await build({
         newContents,
       );
     }
+
+    // Smoke test the built package with node on a machine without a
+    // napi binding (the out dir never contains the platform packages):
+    // importing must succeed, the in-memory implementation must work,
+    // and the napi path must fail with the binding load error attached
+    // as the cause (#99).
+    console.log("smoke testing the built package");
+    const smokeTest = `
+(async () => {
+  const assert = require('assert').strict;
+  const { pathToFileURL } = require('url');
+  const esm = await import(pathToFileURL(${
+      JSON.stringify(join(outDir, "esm", "npm.js"))
+    }));
+  const cjs = require(${JSON.stringify(join(outDir, "script", "npm.js"))});
+  for (const { openKv } of [esm, cjs]) {
+    const kv = await openKv(undefined, { implementation: 'in-memory' });
+    await kv.set(['a'], 1);
+    assert.equal((await kv.get(['a'])).value, 1);
+    kv.close();
+    await assert.rejects(
+      () => openKv('./smoke-test.sqlite'),
+      (e) =>
+        /No default napi interface/.test(e.message) &&
+        e.cause !== undefined,
+    );
+  }
+  console.log('smoke test ok');
+})().catch((e) => {
+  console.error(e);
+  process.exit(1);
+});
+`;
+    await run({ command: "node", args: ["-e", smokeTest] });
   },
 });
 

